flask_common/documents.py

```diff
@@ -288,19 +310,28 @@ def setattr_unchanged(obj, key, val):
 
     # Fetch objects and cache them
     for document_class, fetch_opts in fetch_map.iteritems():
-        qs = document_class.objects.filter(pk__in=fetch_opts['ids']).clear_initial_query()
-
-        # only fetch the requested fields
-        if fetch_opts['fields_to_fetch']:
-            qs = qs.only(*fetch_opts['fields_to_fetch'])
-
-        # update the cache map - either the persistent one with full objects,
-        # or the ephemeral partial cache
-        update_dict = { obj.pk: obj for obj in qs }
-        if fetch_opts['fields_to_fetch'] is None:
-            cache_map[document_class].update(update_dict)
-        else:
-            partial_cache_map[document_class].update(update_dict)
+        cls_filters = extra_filters.get(document_class, {})
+
+        # Fetch objects in batches. Also set the batch size so we don't do
+        # multiple queries per batch.
+        for id_group in grouper(batch_size, list(fetch_opts['ids'])):
+            qs = (document_class.objects.filter(pk__in=id_group, **cls_filters)
+                                        .clear_initial_query())
+
+            # only fetch the requested fields
+            if fetch_opts['fields_to_fetch']:
+                qs = qs.only(*fetch_opts['fields_to_fetch'])
+
+            # We have to apply this at the end, or only() won't work.
+            qs = qs.batch_size(batch_size)
```

Comment:
Why do we need to limit the size here?  Isn't it already limited because `len(id_group)==batch_size`?

Reply:
This would only make a difference for batch sizes over 100, so that we can fetch all the objects in one query. Do you think we should just leave it at the default batch size regardless of how many items the `in__` query contains?

Reply:
I did a little more research but that didn't necessarily clear up how that batch size works for me.  However, after consulting with @jtharpla I think we should remove that line and just rely on the default for that batch size.

Reply:
I'm fine with removing that line, but what's the reasoning?

Reply:
If the default is really `100` (or `101`?) then my concern would be if someone using this code used a batch size of 1,000 or greater we could cause new issues (I think @jtharpla said there is a 16MB limit we need to be careful about also).  So since I don't really understand what is going on underneath all of this I wouldn't increase beyond the default without further research and don't need to make it smaller.

Reply:
While you can raise the default batch size for Mongo queries above 101, Mongo responses are still limited to 16 MB.  So you can't raise the batch size to an arbitrary large number because you might still hit the 16 MB limit depending on the avg doc size.  So I'm thinking it might be better to just leave the queryset batch size at the default and if `batch_size` is larger than this, it just means we need to exhaust the cursor on each id_group.

Reply:
I'm going to slightly reverse my opinion after doing some additional research.  First two thoughts:
* In my opinion using "roundtrips" in the docs is slightly misleading.  It suggests everything could fit into a single "mega" packet and reduce latencies.  No matter what the setting for `batch_size` there are a lot of round trips going on.  Also round trips between backend systems versus across the Internet are obviously different problems to solve.  The real issue is how these setting changes internal Mongo "batching" and query processing.
* It is also important to not focus too much on optimizing this single query but rather keep the whole Mongo cluster in mind.  In my opinion highly concurrent systems work better with a lot of small tasks compared to few large tasks.  Obviously the devil is in the details but I wouldn't necessarily try to optimize this further until I saw a need.

But TIL:
* Pymongo sets the batch size to `0` by default [here](https://github.com/mongodb/mongo-python-driver/blob/4828e426d8058d5a8425a80361f3c6437cd3461a/pymongo/cursor.py#L188)
* Pymongo then sets `numberToReturn` in the query message to this default of `0` [here](https://github.com/mongodb/mongo-python-driver/blob/4828e426d8058d5a8425a80361f3c6437cd3461a/pymongo/cursor.py#L1084)
* Based on the wire protocol [docs](https://docs.mongodb.com/manual/reference/mongodb-wire-protocol/#op-query) a `numberToReturn` of `0` means the server defaults will be used.
* Now it gets interesting
   * For V3.2 the [docs](https://docs.mongodb.com/v3.2/tutorial/iterate-a-cursor/#cursor-batches) say the first batch is 101 documents and then 4 MB batches
  * For V3.4+ the [docs](https://docs.mongodb.com/v3.4/tutorial/iterate-a-cursor/#cursor-batches) the results would potentially just be in 16 MB batches

So, it is probably ok to set `batch_size` to a larger number since only the first batch is limited to 101 documents but I would probably just leave it using the server defaults which would be 16MB once we upgraded to V3.4.  The real question is if they made server changes in V3.4 that justified changing the default to 16MB or did they just realize it was silly to use small batch sizes?

Reply:
> For V3.4+ the docs the results would potentially just be in 16 MB batches

Where does it say that? Your link still states that "find() and aggregate() operations have an initial batch size of 101 documents by default."

Reply:
Looks like I misread that.  Are they just saying the 4MB batch went to 16MB in v3.4+?  Nonetheless, I guess they kept the first batch of 101 even in V4.0.